// ais-stream-decoder/dist/messages/ais-message-4.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ais_message_1 = __importDefault(require("./ais-message"));
const format_1 = __importDefault(require("../format"));
class AisMessage4 extends ais_message_1.default {
    constructor(messageType, channel, bitField) {
        super(messageType, channel, bitField);
        this.year = format_1.default.year(bitField.getInt(38, 14));
        this.month = format_1.default.month(bitField.getInt(52, 4));
        this.day = format_1.default.day(bitField.getInt(56, 5));
        this.hour = format_1.default.hour(bitField.getInt(61, 5));
        this.minute = format_1.default.minute(bitField.getInt(66, 6));
        this.second = format_1.default.second(bitField.getInt(72, 6));
        this.accuracy = bitField.getBoolean(78, 1);
        this.lon = format_1.default.longitude(bitField.getSignedInt(79, 28));
        this.lat = format_1.default.latitude(bitField.getSignedInt(107, 27));
        this.epfd = bitField.getInt(134, 4);
        this.raim = bitField.getBoolean(148, 1);
        this.radio = bitField.getInt(149, 19);
    }
}
exports.default = AisMessage4;
