// ais-stream-decoder/dist/messages/ais-message-18.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ais_message_1 = __importDefault(require("./ais-message"));
const format_1 = __importDefault(require("../format"));
class AisMessage18 extends ais_message_1.default {
    // eslint-disable-next-line max-statements
    constructor(messageType, channel, bitField) {
        super(messageType, channel, bitField);
        this.speedOverGround = format_1.default.speedOverGround(bitField.getInt(46, 10));
        this.accuracy = bitField.getBoolean(56, 1);
        this.lon = format_1.default.longitude(bitField.getSignedInt(57, 28));
        this.lat = format_1.default.latitude(bitField.getSignedInt(85, 27));
        this.courseOverGround = format_1.default.courseOverGround(bitField.getInt(112, 12));
        this.heading = format_1.default.heading(bitField.getInt(124, 9));
        this.utcSecond = bitField.getInt(133, 6);
        this.regional = bitField.getInt(139, 2);
        this.unitFlag = bitField.getBoolean(141, 1);
        this.displayFlag = bitField.getBoolean(142, 1);
        this.dscFlag = bitField.getBoolean(143, 1);
        this.bandFlag = bitField.getBoolean(144, 1);
        this.msg22Flag = bitField.getBoolean(145, 1);
        this.modeFlag = bitField.getBoolean(146, 1);
        this.raim = bitField.getBoolean(147, 1);
        this.radio = bitField.getInt(148, 20);
    }
}
exports.default = AisMessage18;
