// ais-stream-decoder/dist/messages/ais-message-5.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ais_message_1 = __importDefault(require("./ais-message"));
const format_1 = __importDefault(require("../format"));
class AisMessage5 extends ais_message_1.default {
    // eslint-disable-next-line max-statements
    constructor(messageType, channel, bitField) {
        super(messageType, channel, bitField);
        this.aisVersion = bitField.getInt(38, 2);
        this.imo = bitField.getInt(40, 30);
        this.callsign = bitField.getString(70, 42);
        this.name = bitField.getString(112, 120);
        this.typeAndCargo = bitField.getInt(232, 8);
        this.dimBow = bitField.getInt(240, 9);
        this.dimStern = bitField.getInt(249, 9);
        this.dimPort = bitField.getInt(258, 6);
        this.dimStarboard = bitField.getInt(264, 6);
        this.epfd = bitField.getInt(270, 4);
        this.etaMonth = format_1.default.month(bitField.getInt(274, 4));
        this.etaDay = format_1.default.day(bitField.getInt(278, 5));
        this.etaHour = format_1.default.hour(bitField.getInt(283, 5));
        this.etaMinute = format_1.default.minute(bitField.getInt(288, 6));
        this.draught = format_1.default.draught(bitField.getInt(294, 8));
        this.destination = bitField.getString(320, 120);
        this.dte = bitField.getBoolean(422, 1);
    }
}
exports.default = AisMessage5;
