// ais-stream-decoder/dist/messages/ais-message-123.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ais_message_1 = __importDefault(require("./ais-message"));
const format_1 = __importDefault(require("../format"));
class AisMessage123 extends ais_message_1.default {
    constructor(messageType, channel, bitField) {
        super(messageType, channel, bitField);
        this.navStatus = bitField.getInt(38, 4);
        this.rateOfTurn = format_1.default.rateOfTurn(bitField.getSignedInt(42, 8));
        this.speedOverGround = format_1.default.speedOverGround(bitField.getInt(50, 10));
        this.accuracy = bitField.getBoolean(60, 1);
        this.lon = format_1.default.longitude(bitField.getSignedInt(61, 28));
        this.lat = format_1.default.latitude(bitField.getSignedInt(89, 27));
        this.courseOverGround = format_1.default.courseOverGround(bitField.getInt(116, 12));
        this.heading = format_1.default.heading(bitField.getInt(128, 9));
        this.utcSecond = bitField.getInt(137, 6);
        this.specialManoeuvre = bitField.getInt(143, 2);
        this.raim = bitField.getBoolean(148, 1);
        this.radio = bitField.getInt(149, 19);
    }
}
exports.default = AisMessage123;
